classes/completion/custom_completion.php

```diff
@@ -61,50 +80,81 @@ public function get_state(string $rule): int {
                 }
                 // Skip non-target places when there is no condition to visit all places.
-                if ($map->completiontype != 3 && !in_array($place->id, $placestore->targetplaces)) {
+                if ($map->completiontype != self::COMPLETION_WITH_ALL_PLACES && !in_array($place->id, $placestore->targetplaces)) {
                     continue;
                 }
                 if ($place->linkedActivity != null) {
                     if (in_array($place->linkedActivity, $allcms)) {
                         $placecm = $modinfo->get_cm($place->linkedActivity);
                     } else {
                         // No way to fulfill condition.
-                        if ($map->completiontype > 1) {
+                        if ($map->completiontype > self::COMPLETION_WITH_ONE_TARGET) {
                             return COMPLETION_INCOMPLETE;
                         }
                         $placecm = false;
                     }
 
                     if (
                         !$placecm ||
-                        $completion->get_data($placecm, false, $this->userid)->completionstate == COMPLETION_INCOMPLETE
+                        !$this->is_completed($placecm)
                     ) {
                         // No way to fulfill condition.
-                        if ($map->completiontype > 1) {
+                        if ($map->completiontype > self::COMPLETION_WITH_ONE_TARGET) {
                             return COMPLETION_INCOMPLETE;
                         }
                     } else {
                         // We need only one.
                         if (
-                            $map->completiontype == 1 &&
-                            $completion->get_data($placecm, false, $this->userid)->completionstate != COMPLETION_INCOMPLETE
+                            $map->completiontype == self::COMPLETION_WITH_ONE_TARGET &&
+                            $this->is_completed($placecm)
                         ) {
                             return COMPLETION_COMPLETE;
                         }
                     }
                 } else {
                     // No way to fulfill condition.
-                    if ($map->completiontype > 1) {
+                    if ($map->completiontype > self::COMPLETION_WITH_ONE_TARGET) {
                         return COMPLETION_INCOMPLETE;
                     }
                 }
             }
-            if ($map->completiontype == 1) {
+            if ($map->completiontype == self::COMPLETION_WITH_ONE_TARGET) {
                 return COMPLETION_INCOMPLETE;
             } else {
                 return COMPLETION_COMPLETE;
             }
         }
+        return COMPLETION_INCOMPLETE;
+    }
+
+    /**
+     * Checks whether a given course module is completed (either by the user or at least one
+     * of the users of the group, if groupmode is set for the activity).
+     *
+     * @param cm_info $cm course module to check
```

Comment:
```suggestion
     * @param \cm_info $cm course module to check
```
or use import, also in function header